rapids-4-spark-tools/src/main/scala/com/nvidia/spark/rapids/tool/profiling/GenerateDot.scala

```diff
@@ -0,0 +1,208 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.nvidia.spark.rapids.tool.profiling
+
+import java.io.{File, FileWriter}
+import java.util.concurrent.TimeUnit
+
+import org.apache.spark.sql.execution.SparkPlanInfo
+import org.apache.spark.sql.execution.metric.SQLMetricInfo
+
+/**
+ * Generate a DOT graph for one query plan, or showing differences between two query plans.
+ *
+ * Diff mode is intended for comparing query plans that are expected to have the same
+ * structure, such as two different runs of the same query but with different tuning options.
+ *
+ * When running in diff mode, any differences in SQL metrics are shown. Also, if the plan
+ * starts to deviate then the graph will show where the plans deviate and will not recurse
+ * further.
+ *
+ * Graphviz and other tools can be used to generate images from DOT files.
+ *
+ * See https://graphviz.org/pdf/dotguide.pdf for a description of DOT files.
+ */
+object GenerateDot {
+  private val GPU_COLOR = "#76b900" // NVIDIA Green
+  private val CPU_COLOR = "#0071c5"
+  private val TRANSITION_COLOR = "red"
+
+  /**
+   * Generate a query plan visualization in dot format.
+   *
+   * @param plan First query plan and metrics
+   * @param comparisonPlan Optional second query plan and metrics
+   * @param filename Filename to write dot graph to
+   * @param includeCodegen Include WholeStageCodegen and InputAdapter nodes, if true
+   */
+  def generateDotGraph(
+      plan: QueryPlanWithMetrics,
+      comparisonPlan: Option[QueryPlanWithMetrics],
+      dir: File,
+      filename: String,
+      includeCodegen: Boolean): Unit = {
+
+    var nextId = 1
+
+    def isGpuPlan(plan: SparkPlanInfo): Boolean = {
+      plan.nodeName match {
+        case name if name contains "QueryStage" =>
+          plan.children.isEmpty || isGpuPlan(plan.children.head)
+        case name if name == "ReusedExchange" =>
+          plan.children.isEmpty || isGpuPlan(plan.children.head)
+        case name =>
+          name.startsWith("Gpu")
+      }
+    }
+
+    /**
+     * Optionally remove code-gen nodes.
+     */
+    def normalize(plan: SparkPlanInfo): SparkPlanInfo = {
+      if (!includeCodegen
+        && (plan.nodeName.startsWith("WholeStageCodegen")
+        || plan.nodeName.startsWith("InputAdapter"))) {
+        plan.children.head
+      } else {
+        plan
+      }
+    }
+
+    def formatMetric(m: SQLMetricInfo, value: Long): String = {
+      val formatter = java.text.NumberFormat.getIntegerInstance
+      m.metricType match {
+        case "timing" =>
+          val ms = value
+          s"${formatter.format(ms)} ms"
+        case "nsTiming" =>
+          val ms = TimeUnit.NANOSECONDS.toMillis(value)
+          s"${formatter.format(ms)} ms"
+        case _ =>
+          s"${formatter.format(value)}"
+      }
+    }
+
+    /** Recursively graph the operator nodes in the spark plan */
+    def writeGraph(
```

Comment:
consider splitting writeGraph in smaller functions. It looks too long

Reply:
Done

Reply:
I made one small improvement here but don't want to go too far because I can't actually test any of this code in this repo until other PRs are merged in.